pennylane_lightning/lightning_qubit.py

```diff
@@ -131,6 +145,62 @@ def apply_lightning(self, state, operations):
 
         return np.reshape(state_vector, state.shape)
 
+    def adjoint_jacobian(self, tape, starting_state=None, use_device_state=False):
+        if not CPP_BINARY_AVAILABLE:
+            return super().adjoint_jacobian(tape, starting_state, use_device_state)
+
+        if self.shots is not None:
+            warn(
+                "Requested adjoint differentiation to be computed with finite shots."
+                " The derivative is always exact when using the adjoint differentiation method.",
+                UserWarning,
+            )
+
+        if len(tape.trainable_params) == 0:
+            return np.array(0)
+
+        for m in tape.measurements:
+            if m.return_type is not Expectation:
+                raise QuantumFunctionError(
+                    "Adjoint differentiation method does not support"
+                    f" measurement {m.return_type.value}"
+                )
+
+        for op in tape.operations:
+            if op.num_params > 1 and not isinstance(op, Rot):
+                raise QuantumFunctionError(
+                    f"The {op.name} operation is not supported using "
+                    'the "adjoint" differentiation method'
+                )
+
+        # Initialization of state
+        if starting_state is not None:
+            ket = np.ravel(starting_state)
+        else:
+            if not use_device_state:
+                self.reset()
+                self.execute(tape)
+            ket = np.ravel(self._pre_rotated_state)
+
+        adj = AdjointJacobianC128()
+
+        obs_serialized = _serialize_obs(tape, self.wire_map)
+        ops_serialized, use_sp = _serialize_ops(tape, self.wire_map)
+
+        ops_serialized = adj.create_ops_list(*ops_serialized)
+
+        tp_shift = tape.trainable_params if not use_sp else {i - 1 for i in tape.trainable_params}
```

Comment:
Since the state-prep ops aren't passed to the cpp layer, this accounts for a shift in the parametric gates numbers selected by the trainable params list. Once state prep is handled cpp side, we can pass through the appropriate prep ops, and this can vanish.